packages/dialtone-tokens/themes/config.js

```diff
@@ -61,3 +190,341 @@ function _setThemeAttributeOnRoot (theme, brand, rootNode) {
+    );
+  }
+
+  if (typeof brandTheme.brand.css !== 'string') {
+    throw new TypeError(
+      '[Dialtone] initDialtoneTheme: brandTheme.brand.css must be a string containing CSS.',
+    );
+  }
+
+  // Validation: mode must be valid
+  if (mode !== 'light' && mode !== 'dark') {
+    throw new TypeError(
+      `[Dialtone] initDialtoneTheme: mode must be 'light' or 'dark', got '${mode}'.`,
+    );
+  }
+
+  // Validation: rootNode must be a valid element
+  if (!rootNode || (typeof rootNode !== 'object')) {
+    throw new TypeError(
+      '[Dialtone] initDialtoneTheme: rootNode must be an HTMLElement or host element with shadowRoot.',
+    );
+  }
+
+  // Warn if someone passed shadowRoot directly instead of the host
+  if (rootNode instanceof ShadowRoot) {
+    console.warn(
+      '[Dialtone] You passed a ShadowRoot directly to initDialtoneTheme(). ' +
+      'Please pass the host element instead. The function will access shadowRoot automatically.\n' +
+      'Correct: initDialtoneTheme(brand, mode, hostElement)\n' +
+      'Incorrect: initDialtoneTheme(brand, mode, hostElement.shadowRoot)',
+    );
+  }
+
+  if (rootNode?.shadowRoot) {
+    rootNode = rootNode.shadowRoot;
+  }
+
+  // Check for duplicate initialization (idempotency protection with soft warnings)
+  const existing = initializationState.get(rootNode);
+  if (existing) {
+    if (existing.brand === brandTheme.brand.name && existing.mode === mode) {
+      console.warn(
+        `[Dialtone] Theme already initialized for this rootNode with brand '${brandTheme.brand.name}' and mode '${mode}'. ` +
+        'Re-applying the same theme may be unnecessary. ' +
+        'If you need to switch themes dynamically, use setBrand() or setMode() instead of calling initDialtoneTheme() again.',
+      );
+    } else {
+      console.warn(
+        `[Dialtone] Theme already initialized for this rootNode. ` +
+        `Previous: brand='${existing.brand}', mode='${existing.mode}'. ` +
+        `New: brand='${brandTheme.brand.name}', mode='${mode}'. ` +
+        'Re-initializing with different parameters. Consider using setBrand()/setMode() for dynamic switching.',
+      );
+    }
+  }
+
+  // Warn if this looks like an embedded app using the wrong rootNode
+  if (rootNode === document.documentElement && existing) {
+    const hasThemeStyles = document.querySelector('#dialtone-css-core') !== null;
+    if (hasThemeStyles) {
+      console.warn(
+        '[Dialtone] Multiple calls to initDialtoneTheme() detected on document.documentElement. ' +
+        'If you are in an embedded app/micro-frontend, you may need to pass your own container element instead. ' +
+        'Example: initDialtoneTheme(core, brand, mode, myAppContainerElement)',
+      );
+    }
+  }
+
+  // Load core tokens (once per rootNode)
+  _setStyleTag('dialtone-css-core', Core.core, rootNode);
+  coreTokensLoadedByRoot.set(rootNode, true);
```

Comment:
Couldn't this just be a basic boolean, true or false? We never manage more than one root node within a single javascript instance.

Reply:
This week I learned about WeakMap. So...

It previously was a boolean.This update addressed https://github.com/dialpad/dialtone/pull/945#discussion_r2492770118, e.g. micro-fronted and shadow dom increasingly at play.

Before I made that change, we had a true/falsae flag for "are core tokens loaded?" However with multiple containers (Shadow DOM, embedded apps, etc), they'd interfere with each other... first one loads, flag flips to true, second one thinks it's done and skips loading.

WeakMap tracks each separately... "has **_this_** specific element been set up?" instead of "has **_any_** element been set up?" Each gets its own independent tracking.

Switching back to a boolean would bring that bug back. Hope I'm making sense.

Reply:
> Before I made that change, we had a true/falsae flag for "are core tokens loaded?" However with multiple containers (Shadow DOM, embedded apps, etc), they'd interfere with each other... first one loads, flag flips to true, second one thinks it's done and skips loading.

What you'd have in the case of a shadow dom app, is a completely separate instance of this config.js, and the variables within it. coreTokensLoadedByRoot would exist once in the dialpad/ubervoice app, another instance of coreTokensLoadedByRoot would exist in the shadowDom app. The coreTokensLoadedByRoot state is not shared between the two.

Reply:
Updated in https://github.com/dialpad/dialtone/pull/945/commits/49240312b20b087c6da5b4c341a61d8d6ab329f5

Simplified out from WeakMaps. per-rootNode tracking was solving a problem that won't occur. `coreTokensLoadedByRoot` is back to boolean, and `initializationState` as plain object. Simpler, same functionality.